internal/core/application/walletunlocker_service.go

```diff
@@ -0,0 +1,950 @@
+		chRes chan *InitWalletReply,
+		chErr chan error,
+	)
+	UnlockWallet(
+		ctx context.Context,
+		passphrase string,
+	) error
+	ChangePassword(
+		ctx context.Context,
+		currentPassphrase string,
+		newPassphrase string,
+	) error
+	IsReady(ctx context.Context) bool
+	PassphraseChan() chan PassphraseMsg
+	ReadyChan() chan bool
+}
+
+type walletUnlockerService struct {
+	repoManager        ports.RepoManager
+	explorerService    explorer.Service
+	blockchainListener BlockchainListener
+	walletInitialized  bool
+	walletIsSyncing    bool
+	network            *network.Network
+	marketFee          int64
+	marketBaseAsset    string
+
+	lock      *sync.RWMutex
+	pwChan    chan PassphraseMsg
+	readyChan chan bool
+}
+
+func NewWalletUnlockerService(
+	repoManager ports.RepoManager,
+	explorerService explorer.Service,
+	blockchainListener BlockchainListener,
+	net *network.Network,
+	marketFee int64,
+	marketBaseAsset string,
+) WalletUnlockerService {
+	return newWalletUnlockerService(
+		repoManager,
+		explorerService,
+		blockchainListener,
+		net,
+		marketFee,
+		marketBaseAsset,
+	)
+}
+
+func newWalletUnlockerService(
+	repoManager ports.RepoManager,
+	explorerService explorer.Service,
+	blockchainListener BlockchainListener,
+	net *network.Network,
+	marketFee int64,
+	marketBaseAsset string,
+) *walletUnlockerService {
+	w := &walletUnlockerService{
+		repoManager:        repoManager,
+		explorerService:    explorerService,
+		blockchainListener: blockchainListener,
+		network:            net,
+		marketFee:          marketFee,
+		marketBaseAsset:    marketBaseAsset,
+		lock:               &sync.RWMutex{},
+		pwChan:             make(chan PassphraseMsg, 1),
+		readyChan:          make(chan bool, 1),
+	}
+	// to understand if the service has an already initialized wallet we check
+	// if the inner vaultRepo is able to return a Vault without passing mnemonic
+	// and passphrase. If it does, it means it's been retrieved from storage,
+	// therefore we let the crawler to start watch all derived addresses and mark
+	// the wallet as initialized
+	if vault, err := w.repoManager.VaultRepository().GetOrCreateVault(
+		context.Background(), nil, "", nil,
+	); err == nil {
+		go func() {
+			log.Info("Restoring internal wallet's utxo set. This could take a while...")
+
+			info := vault.AllDerivedAddressesInfo()
+			if err := fetchAndAddUnspents(
+				w.explorerService,
+				w.repoManager.UnspentRepository(),
+				w.blockchainListener,
+				info,
+			); err != nil {
+				log.Infof("Failed for reason: %s", err)
+				w.setInitialized(false)
+				return
+			}
+			log.Info("Done")
+			w.setInitialized(true)
+		}()
+	}
+	return w
+}
+
+func (w *walletUnlockerService) GenSeed(ctx context.Context) ([]string, error) {
+	mnemonic, err := wallet.NewMnemonic(wallet.NewMnemonicOpts{EntropySize: 256})
+	if err != nil {
+		return nil, err
+	}
+	return mnemonic, nil
+}
+
+func (w *walletUnlockerService) IsReady(_ context.Context) bool {
+	return w.isInitialized()
```

Comment:
is it isInitiazed used to know if the seed is there? wouldnt be better adding another state like `isRestored`

so we check if `isInitialized` && `isRestored`

Reply:
`isInitialized` expresses the condition of a wallet that has been created/loaded from db and its UTXO set is fully synced.
The service knows if a wallet has been created by attempting to get the vault form the db and checking if it's nil eventually.